### ISSUE TITLE
Sites using location-bar or similar breaks in unblocker

### ISSUE THREAD

**Issue by emilhem** [nfriedly/node-unblocker#47]
So there is this technology that more and more sites are starting to use that breaks the way that unblocker handles paths. (https://github.com/KidkArolis/location-bar)

The technology enables JavaScript to change the complete path (not including domains) for the client without reloading the browser.

I noticed this problem when trying to make DynaMeds page work in my project Proxive.

Other examples of this technology in use:
Discourse: http://try.discourse.org/
LinkedIn Pulse: https://www.linkedin.com/pulse/how-those-totally-useless-meetings-can-make-break-your-jack-welch
Ember.js uses this technology in its routers.

To fix this it would require a complete rewrite of unblocker so it instead uses the domain for the domain of the page visited like this:

`http://google.com.example.com/path?query=query#hash` instead of the current `http://example.com/proxy/http://google.com/path?query=query#hash`

The biggest issue I see is how to detect wheter the site is https or not. Maybe a solution would be prepending the domain with something like `h-t-t-p-s` so that the domain would look like `h-t-t-p-s.google.com.example.com`.

You can close this issue if you feel that this will not/cannot be part of this project. Maybe we can cooperate on a separate project/branch that enables the above? (I'm planning on working on this very soon since my job needs to stay on top of the issue so please respond as soon as possible if you're interested in cooperating.)

**Comment by emilhem** [comment]
I can add that I'm making progress on this! :)

**Comment by nfriedly** [comment]
Yea, I've thought about adding a subdomain mode a few times. Part of the problem is that it would be much more difficult to test and to set up a private instance, as you then need a proper DNS server. I can set up a `*.localhost.nodeunblocker.com` DNS entry that always points to localhost to make testing easier, but that doesn't help for users who want to run their own copy. It might still be worth it, but it's a major change to functionality. 

It would make cookie processing much simpler.

As for http/https - I think the simple solution is to just make the server support both and match the requested protocol. But again, this makes setup more complex. (Although, if you don't mind a self-signed certificate, I can include one with the project.)

There is an alternate option of injecting JS that wraps [pushState](https://developer.mozilla.org/en-US/docs/Web/Guide/API/DOM/Manipulating_the_browser_history#Adding_and_modifying_history_entries)  - I've been thinking about this to work around land other APIs that affect or depend on the URL. I've been thinking about this to work around login issues that depend on postMessage, so it's worth considering. pushState would probably be easier to wrap than postMessage since it only deals with the path.

**Comment by emilhem** [comment]
A potential fix that's not good but will work in most cases would to make unblocker prepend a script tag with the following.
```JavaScript
delete window.history
delete history // It might do it to just do the above.
```
Since many tools today checks whether `window.history` or `history` exists and fall back on hashchange instead it could work for a while until all browsers support `popState`

**Comment by nfriedly** [comment]
Heh, yea, that would probably work too, at least for a little while. Although prettymuch all browsers already support it (http://caniuse.com/#feat=history) so I can certainly imagine sites that just assume it will exist. 

I tried that with postMessage, and a lot of stuff was still broken :/

**Comment by emilhem** [comment]
Hehe, tried it and it caused issues of course... :/

I will attempt to make a domain version like I mentioned above. I'll keep you posted.

**Comment by nfriedly** [comment]
Cool, sounds good

**Comment by emilhem** [comment]
I'm making great progress!

Should I make a pull request with the change as an option in config when I've tested it more?

**Comment by nfriedly** [comment]
Yea, please do

**Comment by emilhem** [comment]
Almost a year later... I've partially stopped working on the company where I develop this and I will quit in the middle of august. I'm working until then 100% on the proxy. I have pushed some changes that I made a long time ago with some tests to my fork.

**Comment by nfriedly** [comment]
Awesome, and no worries about the delay.  I'm at work right now but I'll
check out your stuff and respond as time allows.

On Mon, Jul 11, 2016 at 11:08 AM, Emil Hemdal <notifications@github.com>
wrote:

> Almost a year later... I've partially stopped working on the company where
> I develop this and I will quit in the middle of august. I'm working until
> then 100% on the proxy. I have pushed some changes that I made a long time
> ago with some tests to my fork.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nfriedly/node-unblocker/issues/47#issuecomment-231763339>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHBIID-_Wu8soFQn8rT5NIX4cAXM55Fks5qUlxwgaJpZM4FmsbD>
> .
>



-- 
Nathan Friedly
Minister, JavaScript Ninjaneer
http://nfriedly.com

**Comment by nfriedly** [comment]
I know this is a super-old ticket, but I fixed this issue in v2.3.0 in the process of trying to make youtube work. 

The proxy now injects some JavaScript that wraps `history.pushState()` and `history.replaceState()` to fixup the URLs before passing the call along to the browser. 

So if the website's code calls 
```js
history.pushState({}, 'title', '/foo/bar')
```
the proxy code will intercept that and rewrite it to
```js
history.pushState({}, 'title', '/proxy/foo/bar')
```
(or whatever you have prefix configured to)
